### PR TITLE
[r] Add SOMACollection class

### PR DESCRIPTION
**Pull request by aaronwolen**
This provides an initial implementation of the `SOMACollection` class. 

## Details 
Much of the logic for managing group members is implemented in the new `TileDBGroup` class. `SOMACollectionBase` is a subclass of `TileDBGroup` and adds SOMA-specific functionality such as automatic writing of SOMA metadata on creation and construction of `SOMA` objects when retrieving group members. 

Currently, only `SOMACollection` inherits from `SOMACollectionBase`. However, in future PRs it will also be used for the implementation of `SOMAExperiment` and `SOMAMeasurement`.

This is also adds a new `SOMArrayBase` class that is a subclass of `TileDBArray` and adds SOMA-specific functionality to `SOMADataFrame`, `SOMADenseNDArray`, and `SOMASparseNDArray`.

## Other changes:

- Fixes print method for `TileDBObject`-based classes
- Add missing `fs` import
- Add R CI badge to README

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <a href="https://tiledb.com"><img src="https://github.com/TileDB-Inc/TileDB/raw/dev/doc/source/_static/tiledb-logo_color_no_margin_@4x.png" alt="TileDB logo" width="400"></a>
 
-[![TileDB-SOMA CI](https://github.com/single-cell-data/TileDB-SOMA/actions/workflows/ci.yml/badge.svg)](https://github.com/single-cell-data/TileDB-SOMA/actions/workflows/ci.yml) [![codecov](https://codecov.io/github/single-cell-data/TileDB-SOMA/branch/main/graph/badge.svg)](https://codecov.io/github/single-cell-data/TileDB-SOMA)
+[![TileDB-SOMA CI](https://github.com/single-cell-data/TileDB-SOMA/actions/workflows/ci.yml/badge.svg)](https://github.com/single-cell-data/TileDB-SOMA/actions/workflows/ci.yml) [![TileDB-SOMA R CI](https://github.com/single-cell-data/TileDB-SOMA/actions/workflows/r-ci.yml/badge.svg)](https://github.com/single-cell-data/TileDB-SOMA/actions/workflows/r-ci.yml) [![codecov](https://codecov.io/github/single-cell-data/TileDB-SOMA/branch/main/graph/badge.svg)](https://codecov.io/github/single-cell-data/TileDB-SOMA)
```
Comment (eddelbuettel) [inline README.md:3]: It's easiest to insert a line break. Could you add one before the codecov entry?

File: apis/r/R/SOMACollectionBase.R [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        is_scalar_character(uri),
+        is_scalar_character(type)
+      )
+
+      # We have to use the appropriate TileDB base class to read the soma_type
+      # from the object's metadata so we knowh which SOMA class to instantiate
+      tiledb_constructor <- switch(type,
+        ARRAY = TileDBArray$new,
+        GROUP = TileDBGroup$new,
+        stop(sprintf("Unknown member TileDB type: %s", type))
```
Comment (eddelbuettel) [inline apis/r/R/SOMACollectionBase.R:90]: Maybe add `, call. = FALSE` at end of `stop()` ?

File: apis/r/R/SOMACollectionBase.R [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+      tiledb_object <- tiledb_constructor(uri, self$ctx, self$platform_config)
+      soma_type <- tiledb_object$get_metadata(SOMA_OBJECT_TYPE_METADATA_KEY)
+
+      soma_constructor <- switch(soma_type,
+        SOMADataFrame = SOMADataFrame$new,
+        SOMADenseNdArray = SOMADenseNdArray$new,
+        SOMASparseNdArray = SOMASparseNdArray$new,
+        SOMACollection = SOMACollection$new,
+        stop(sprintf("Unknown member SOMA type: %s", soma_type))
```
Comment (eddelbuettel) [inline apis/r/R/SOMACollectionBase.R:101]: Maybe add `, call. = FALSE` at end of `stop()` ?

**Review by eddelbuettel — COMMENTED** [review]
Looks good at first quick glance, I'll try to install it in a bit and give it a spin.

**Review by eddelbuettel — APPROVED** [review]
Looks good to me.  If you have a sec maybe address the one nit else we do it later

File: apis/r/R/SOMAArrayBase.R [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+SOMAArrayBase <- R6::R6Class(
+  classname = "SOMAArrayBase",
+  inherit = TileDBArray,
+
+  active = list(
+    #' @field soma_type Retrieve the SOMA object type.
+    soma_type = function(value) {
+      if (!missing(value)) {
+        stop("`soma_type` is a read-only field", call. = FALSE)
+      }
```
Comment (eddelbuettel) [inline apis/r/R/SOMAArrayBase.R:14]: Simpler?  `stopifnot("`soma_type` is a read-only field" = missing(value)` ?  Also in these quotes simple apostrophe may be nicer than backtick.

File: apis/r/R/SOMACollectionBase.R [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+  active = list(
+    #' @field soma_type Retrieve the SOMA object type.
+    soma_type = function(value) {
+      if (!missing(value)) {
+        stop("`soma_type` is a read-only field", call. = FALSE)
+      }
+      if (is.null(private$soma_type_cache)) {
+        private$update_soma_type_cache()
+      }
```
Comment (eddelbuettel) [inline apis/r/R/SOMACollectionBase.R:56]: See above. Collapse six lines into two using `stopifnot()` ?

File: apis/r/R/TileDBObject.R [hunk 6]
```diff
@@ -7,31 +7,22 @@
...
+    #' @field platform_config Optional platform configuration
+    platform_config = NULL,
+    #' @field ctx Optional TileDB context
     ctx = NULL,
 
     #' @description Create a new TileDB object.
     #' @param uri URI for the TileDB object
     #' @param verbose Print status messages
-    #' @param config optional configuration
+    #' @param platform_config Optional platform configuration
```
Comment (eddelbuettel) [inline apis/r/R/TileDBObject.R:18]: Should we detail what we expect here?   If we can do without nesting a named character vectors works (in the context of TileDB Core config where right-hand side values are all character).

Comment (aaronwolen) [reply apis/r/R/TileDBObject.R:18]: This is a placeholder at the moment. We can round out the platform_config functionality in a future PR.

**Comment by aaronwolen** [comment]
Thanks for feedback, @eddelbuettel. I incorporated all suggestions in the last two commits.

**Review by eddelbuettel — APPROVED** [review]
:+1:

**Review by johnkerl — APPROVED** [review]

**Pull request merged**